transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java

```diff
@@ -0,0 +1,217 @@
+                    "This metadata was not found under " + existingMetadataPath);
+            System.err.println("Please create this file with the following content: ");
+            System.err.println(getMetadataJsonString(handlerMetadata));
+            Assertions.fail();
+        }
+
+        String existingMetadataJson = "";
+        try {
+            existingMetadataJson = Files.readString(existingMetadataPath);
+        } catch (IOException e) {
+            Assertions.fail("Failed to open the native-image metadata file at: " + existingMetadataPath, e);
+        }
+        List<HandlerMetadata> existingMetadata = gson.fromJson(existingMetadataJson, HANDLER_METADATA_LIST_TYPE);
+
+        Set<HandlerMetadata> newMetadata = new HashSet<>(handlerMetadata);
+        newMetadata.removeAll(existingMetadata);
+
+        Set<HandlerMetadata> removedMetadata = new HashSet<>(existingMetadata);
+        removedMetadata.removeAll(handlerMetadata);
+
+        if (!newMetadata.isEmpty() || !removedMetadata.isEmpty()) {
+            System.err.println("In the native-image handler metadata file at " + existingMetadataPath);
+
+            if (!newMetadata.isEmpty()) {
+                System.err.println("The following new metadata must be added:\n");
+                System.err.println(getMetadataJsonString(newMetadata));
+                System.err.println();
+            }
+            if (!removedMetadata.isEmpty()) {
+                System.err.println("The following metadata must be removed:\n");
+                System.err.println(getMetadataJsonString(removedMetadata));
+                System.err.println();
+            }
+
+            System.err.println("Expected metadata file contents:\n");
+            System.err.println(getMetadataJsonString(handlerMetadata));
+            Assertions.fail();
+        }
+    }
+
+    private static Set<Class<? extends ChannelHandler>> findChannelHandlerSubclasses(String... packageNames) {
+        Reflections reflections = new Reflections(
+                new ConfigurationBuilder()
+                        .forPackages(packageNames)
+                        .filterInputsBy(s -> !s.contains("Test")));
+
+        Set<Class<? extends ChannelHandler>> classes = reflections.getSubTypesOf(ChannelHandler.class);
+        classes = classes.stream()
+                .filter(ChannelHandlerMetadataUtil::isTestClass)
+                .filter(clazz -> Stream.of(packageNames).anyMatch(name -> clazz.getName().startsWith(name)))
+                .collect(Collectors.toSet());
+        return classes;
+    }
+
+    private static boolean isTestClass(Class<? extends ChannelHandler> clazz) {
+        String[] parts = clazz.getName().split("\\.");
+        if (parts.length > 0) {
+            URL classFile = clazz.getResource(parts[parts.length - 1] + ".class");
+            if (classFile != null) {
+                return !classFile.toString().contains(File.separator + "test-classes" + File.separator);
```

Comment:
This one bugged my mind a bit - is there a smarter way to find out if a class originates from the test sources? It would be nice to have all of the test classes filtered automatically, and this seems to work, but it also seems a bit ugly